<a href="https://colab.research.google.com/github/Dimas0824/Machine_Learning/blob/main/Jobsheet_9/ML_WEEK10.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 🥚 Tugas 2



## Buatlah model klasfikasi Multinomial Naive Bayes dengan ketentuan,



### 1. Menggunakan data spam.csv

In [66]:
# Import library

import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Load dataset
spam = pd.read_csv('/content/drive/MyDrive/Machine Learning/DATASET/spam.csv', encoding='latin-1')

# Hapus kolom yang tidak digunakan
spam = spam.drop(spam.iloc[:, 2:], axis=1)

# Ganti nama kolom agar mudah dipahami
spam.columns = ['Label', 'SMS']

# Tampilkan beberapa baris pertama
print(spam.head())

  Label                                                SMS
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...


### 2. Preprocessing Data

In [67]:
# Encode label: spam = 1, ham = 0
spam['Label'] = spam['Label'].map({'spam': 1, 'ham': 0})

# Pisahkan fitur (X) dan label (y)
X = spam['SMS'].values
y = spam['Label'].values

### 3. Split Data (Training & Testing)

In [68]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

### 4. Ekstraksi Fitur (CountVectorizer dengan Stop Words)

In [69]:
# Inisialisasi CountVectorizer dengan stop words
vectorizer = CountVectorizer(stop_words='english')

# Fit dan transform data training
X_train_vec = vectorizer.fit_transform(X_train)

# Transform data testing (tanpa fit untuk menghindari data leakage)
X_test_vec = vectorizer.transform(X_test)

# Cek jumlah fitur
print(f"Jumlah fitur: {len(vectorizer.get_feature_names_out())}")


Jumlah fitur: 7472


### 5. Training Model Multinomial Naive Bayes

In [70]:
# Inisialisasi model
model = MultinomialNB()

# Latih model dengan data training
model.fit(X_train_vec, y_train)

MultinomialNB()

### 6. Prediksi & Evaluasi Model

In [71]:
# Prediksi data training dan testing
y_pred_train = model.predict(X_train_vec)
y_pred_test = model.predict(X_test_vec)

# Hitung akurasi
acc_train = accuracy_score(y_train, y_pred_train)
acc_test = accuracy_score(y_test, y_pred_test)

# Cetak hasil evaluasi
print("HASIL EVALUASI MODEL")
print(f"Akurasi Data Train: {acc_train:.4f}")
print(f"Akurasi Data Test : {acc_test:.4f}")
print("\nClassification Report:")
print(classification_report(y_test, y_pred_test, target_names=['Ham', 'Spam']))
print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_test))

HASIL EVALUASI MODEL
Akurasi Data Train: 0.9946
Akurasi Data Test : 0.9839

Classification Report:
              precision    recall  f1-score   support

         Ham       0.99      0.99      0.99       965
        Spam       0.96      0.92      0.94       150

    accuracy                           0.98      1115
   macro avg       0.97      0.96      0.96      1115
weighted avg       0.98      0.98      0.98      1115


Confusion Matrix:
[[959   6]
 [ 12 138]]


## Buatlah model klasfikasi Multinomial Naive Bayes dengan ketentuan,

###   2. Fitur TF-IDF dengan mengaktifkan stop_words


In [72]:
from sklearn.feature_extraction.text import TfidfVectorizer

# Inisialisasi TF-IDF Vectorizer dengan stop words
tfidf = TfidfVectorizer(stop_words='english')

# Fit dan transform data training
X_train_tfidf = tfidf.fit_transform(X_train)

# Transform data testing (tanpa fit)
X_test_tfidf = tfidf.transform(X_test)

print(f"Jumlah fitur TF-IDF: {len(tfidf.get_feature_names_out())}")

Jumlah fitur TF-IDF: 7472


#### Training Model

In [73]:
# Inisialisasi model
model_tfidf = MultinomialNB()

# Training
model_tfidf.fit(X_train_tfidf, y_train)


MultinomialNB()

#### Evaluasi Model TF-IDF

In [74]:
# Prediksi data training & testing
y_pred_train_tfidf = model_tfidf.predict(X_train_tfidf)
y_pred_test_tfidf = model_tfidf.predict(X_test_tfidf)

# Hitung akurasi
acc_train_tfidf = accuracy_score(y_train, y_pred_train_tfidf)
acc_test_tfidf = accuracy_score(y_test, y_pred_test_tfidf)

print("HASIL EVALUASI MODEL (TF-IDF)")
print(f"Akurasi Data Train: {acc_train_tfidf:.4f}")
print(f"Akurasi Data Test : {acc_test_tfidf:.4f}")
print("\nClassification Report:")
print(classification_report(y_test, y_pred_test_tfidf, target_names=['Ham', 'Spam']))
print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_test_tfidf))


HASIL EVALUASI MODEL (TF-IDF)
Akurasi Data Train: 0.9838
Akurasi Data Test : 0.9668

Classification Report:
              precision    recall  f1-score   support

         Ham       0.96      1.00      0.98       965
        Spam       1.00      0.75      0.86       150

    accuracy                           0.97      1115
   macro avg       0.98      0.88      0.92      1115
weighted avg       0.97      0.97      0.96      1115


Confusion Matrix:
[[965   0]
 [ 37 113]]


###   3. Evaluasi hasilnya dan bandingkan dengan hasil pada Tugas no 2.

In [75]:
from tabulate import tabulate

# Data perbandingan
comparison = [
    ["CountVectorizer", acc_test],
    ["TF-IDF", acc_test_tfidf]
]

# Format tabel
headers = ["Metode", "Akurasi"]
print("\n PERBANDINGAN ANTARA COUNT VECTORIZER DENGAN TFIDF")
print()
print(tabulate(comparison, headers=headers, floatfmt=".4f", tablefmt="grid"))


 PERBANDINGAN ANTARA COUNT VECTORIZER DENGAN TFIDF

+-----------------+-----------+
| Metode          |   Akurasi |
+=================+===========+
| CountVectorizer |    0.9839 |
+-----------------+-----------+
| TF-IDF          |    0.9668 |
+-----------------+-----------+


Dari hasil diatas, dapat dilihat bahwa model dengan CountVectorizer menghasilkan akurasi lebih tinggi (98.39%) dibandingkan model TF-IDF (96.68%). Hal ini karena kemungkinan pada dataset SPAM pola kata dalam pesan spam cenderung repetitif dan dapat dideteksi dengan lebih baik melalui frekuensi kemunculan kata, bukan bobot seperti pada TF-IDF.

### 4. Berikan kesimpulan fitur mana yang terbaik pada kasus data spam.csv

jawab: Berdasarkan hasil evaluasi, metode **CountVectorizer** menunjukkan performa terbaik dalam klasifikasi pesan spam dibandingkan dengan **TF-IDF**, baik dari segi akurasi maupun relevansi fitur yang dipilih.

| Metode           | Jumlah Fitur Awal |
|------------------|-------------------|
| CountVectorizer  | 7472              |
| TF-IDF           | 7472              |

#### 20 Fitur Terpilih
**CountVectorizer**: ['00', '000', '02', '03', '04', '06', '0800', '08000839402', '08000930705', '0870', '08707509020', '08712300220', '08712460324', '08718720201', '10', '100', '1000', '10p', '11', '11mths']

**TF-IDF**: ['00', '000', '02', '03', '04', '06', '0800', '08000839402', '08000930705', '08001950382', '08002986906', '0870', '08702840625', '08707509020', '08712300220', '08712460324', '08718720201', '09050090044', '10', '100']

Fitur diatas terpilih sebagai fitur terbaik dalam anova test karena dalam feature selection ANOVA menghitung nilai F-statistic untuk setiap fitur. Fitur dengan variasi nilai yang besar antar kelas (misalnya kata yang jauh lebih sering muncul di spam dibanding di ham) akan mendapat skor F yang tinggi. Dengan kata lain, fitur terbaik itu sering muncul dalam dataset.

#### Perbandingan Akurasi
| Metode          | Akurasi  |
|-----------------|----------|
| CountVectorizer | 0.9507   |
| TF-IDF          | 0.8771   |

**Kesimpulan**:  
Metode terbaik untuk kasus klasifikasi spam pada dataset `spam.csv` adalah **CountVectorizer**, karena menghasilkan akurasi yang lebih tinggi dan fitur yang lebih relevan terhadap pola pesan spam.


### Load dan split data

In [85]:
df = pd.read_csv('/content/drive/MyDrive/Machine Learning/DATASET/spam.csv', encoding='latin-1')

# Hapus kolom tidak relevan
df = df.drop(df.iloc[:, 2:], axis=1)
df.columns = ['Label', 'SMS']

# Encode label: spam = 1, ham = 0
df['Label'] = df['Label'].map({'spam': 1, 'ham': 0})

# Split data train dan test
X_train, X_test, y_train, y_test = train_test_split(
    df['SMS'], df['Label'], test_size=0.2, random_state=42
)

### Pemilihan Fitur (menggunakan anova test)

In [86]:
vectorizers = {
    "CountVectorizer": CountVectorizer(stop_words='english'),
    "TF-IDF": TfidfVectorizer(stop_words='english')
}

results = {}

### Train ulang model menggunakan Fitur terpilih

In [95]:
for name, vectorizer in vectorizers.items():
    print(f"\nMetode: {name}")

    # Transformasi data
    X_train_vec = vectorizer.fit_transform(X_train)
    X_test_vec = vectorizer.transform(X_test)
    print(f"Jumlah fitur awal: {X_train_vec.shape[1]}")

    # Seleksi fitur dengan ANOVA
    selector = SelectKBest(score_func=f_classif, k=k_best)
    X_train_sel = selector.fit_transform(X_train_vec, y_train)
    X_test_sel = selector.transform(X_test_vec)
    best_features = vectorizer.get_feature_names_out()[selector.get_support()]
    print("20 fitur terpilih:")
    print(best_features[:20])

    # Training dan evaluasi
    model = MultinomialNB()
    model.fit(X_train_sel, y_train)
    y_pred = model.predict(X_test_sel)
    acc = accuracy_score(y_test, y_pred)

    # Simpan hasil
    results[name] = {
        "accuracy": acc,
        "features": best_features,
        "report": classification_report(y_test, y_pred, target_names=['Ham', 'Spam']),
        "confusion": confusion_matrix(y_test, y_pred)
    }

    # Tampilkan evaluasi
    print(f"Akurasi: {acc:.4f}")
    print("Classification Report:")
    print(results[name]["report"])


Metode: CountVectorizer
Jumlah fitur awal: 7472
20 fitur terpilih:
['00' '000' '02' '03' '04' '06' '0800' '08000839402' '08000930705' '0870'
 '08707509020' '08712300220' '08712460324' '08718720201' '10' '100' '1000'
 '10p' '11' '11mths']
Akurasi: 0.9507
Classification Report:
              precision    recall  f1-score   support

         Ham       0.95      1.00      0.97       965
        Spam       0.97      0.65      0.78       150

    accuracy                           0.95      1115
   macro avg       0.96      0.83      0.88      1115
weighted avg       0.95      0.95      0.95      1115


Metode: TF-IDF
Jumlah fitur awal: 7472
20 fitur terpilih:
['00' '000' '02' '03' '04' '06' '0800' '08000839402' '08000930705'
 '08001950382' '08002986906' '0870' '08702840625' '08707509020'
 '08712300220' '08712460324' '08718720201' '09050090044' '10' '100']
Akurasi: 0.8771
Classification Report:
              precision    recall  f1-score   support

         Ham       0.88      1.00      0.9

### Evaluasi dan Perbandingan kedua model

In [96]:
# Siapkan data akurasi
accuracy_table = [[name, result["accuracy"]] for name, result in results.items()]

# Tampilkan tabel
print("\nPerbandingan Akurasi")
print(tabulate(accuracy_table, headers=["Metode", "Akurasi"], floatfmt=".4f", tablefmt="grid"))

# Tentukan metode terbaik
best_method = max(results, key=lambda k: results[k]['accuracy'])
print(f"\nMetode terbaik untuk dataset spam.csv adalah: {best_method}")


Perbandingan Akurasi
+-----------------+-----------+
| Metode          |   Akurasi |
+=================+===========+
| CountVectorizer |    0.9507 |
+-----------------+-----------+
| TF-IDF          |    0.8771 |
+-----------------+-----------+

Metode terbaik untuk dataset spam.csv adalah: CountVectorizer
